### PR TITLE
[Docs] MessagePack IDL, Pydantic Support, and Attribute Access

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Docs link
- dataclass: https://flyte--6022.org.readthedocs.build/en/6022/user_guide/data_types_and_io/dataclass.html
- pydantic basemodel: https://flyte--6022.org.readthedocs.build/en/6022/user_guide/data_types_and_io/pydantic_basemodel.html
- attribute access: https://flyte--6022.org.readthedocs.build/en/6022/user_guide/data_types_and_io/accessing_attributes.html

## Tracking issue
https://github.com/flyteorg/flyte/issues/5318

### CODE REVIEW

File: docs/user_guide/data_types_and_io/pydantic.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+(pydantic)=
+
+# Pydantic BaseModel
+
+```{eval-rst}
+.. tags:: Basic
+```
+
+When you've multiple values that you want to send across Flyte entities, and you want them to have, you can use a `pydantic.BaseModel`.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic.md:9]: Not sure I understand this line

Comment (Future-Outlier) [reply docs/user_guide/data_types_and_io/pydantic.md:9]: changed

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 2]
```diff
@@ -12,7 +12,18 @@ Flytekit uses the [Mashumaro library](https://github.com/Fatal1ty/mashumaro)
 to serialize and deserialize dataclasses.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:12]:
```suggestion
to serialize and deserialize dataclasses. With the 1.14 release, `flytekit` adopted `MessagePack` as the serialization format for dataclasses, overcoming  a major limitation of serialization into a JSON string within a Protobuf `struct` datatype, like the previous versions do: to store `int` types, Protobuf's `struct` converts them to `float`, forcing users to write boilerplate code to workaround this issue. By default, `flytekit >= 1.14` will produce `msgpack` bytes literals when serializing dataclasses.

:::{important}

If you're serializing dataclasses using `flytekit` version >= v1.14.0 and you want to produce Protobuf `struct literal` instead, you can set environment variable `FLYTE_USE_OLD_DC_FORMAT` to `true`.
```
The reason I suggest the following is to make it a bit more digestable for users and prevent them from having to read the RFC to understand what has changed and why

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
+(pydantic_basemodel)=
+
+# Pydantic BaseModel
+
+```{eval-rst}
+.. tags:: Basic
+```
+
+When you have multiple values that you want to send across Flyte entities, and you want them to have, you can use a `pydantic.BaseModel`.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:9]:
```suggestion
`flytekit` version >=1.14 supports natively the `JSON` format that Pydantic `BaseModel` produces,  enhancing the interoperability of Pydantic schemas with the Flyte type system.
```

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+:::{important}
+Pydantic BaseModel V2 only works when you are using flytekit version >= v1.14.0.
+:::
+
+:::{important}
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:15]:
```suggestion
With the 1.14 release, `flytekit` adopted `MessagePack` as the serialization format for Pydantic `BaseModel`, overcoming a major limitation of serialization into a JSON string within a Protobuf `struct` datatype like the previous versions do: to store `int` types, Protobuf's `struct` converts them to `float`, forcing users to write boilerplate code to workaround this issue. By default, `flytekit >= 1.14` will produce `msgpack` bytes literals when serializing dataclasses, preserving the types defined in your `BaseModel` class.
:::{important}
If you're serializing dataclasses using `flytekit` version >= v1.14.0 and you want to produce Protobuf `struct literal` instead, you can set environment variable `FLYTE_USE_OLD_DC_FORMAT` to `true`.
```

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+If you're using Flytekit version >= v1.14.0 and you want to produce protobuf struct literal for Pydantic BaseModels,
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:16]:
```suggestion

```

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+you can set environment variable  `FLYTE_USE_OLD_DC_FORMAT` to `true`.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:17]:
```suggestion
```

**Comment by davidmirror-ops** [comment]
@Future-Outlier left some suggestions to improve clarity. I hope it helps

**Comment by Future-Outlier** [comment]
> @Future-Outlier left some suggestions to improve clarity. I hope it helps

thank you so much David, will finish this today

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 7]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
 Flytekit uses the [Mashumaro library](https://github.com/Fatal1ty/mashumaro)
 to serialize and deserialize dataclasses.
 
+With the 1.14 release, `flytekit` adopted `MessagePack` as the 
+serialization format for dataclasses, overcoming  a major limitation of serialization into a JSON string within a Protobuf `struct` datatype, like the previous versions do:
+
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:16]:
```suggestion
```

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 8]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
...
+to store `int` types, Protobuf's `struct` converts them to `float`, forcing users to write boilerplate code to work around this issue.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:17]: no need to insert a new line

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 9]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
...
+
+By default, `flytekit >= 1.14` will produce `msgpack` bytes literals when serializing dataclasses.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:19]:
```suggestion
```

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 10]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
...
+
+:::{important}
+
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:22]:
```suggestion

By default, `flytekit >= 1.14` will produce `msgpack` bytes literals when serializing dataclasses.
```

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 11]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
...
+If you're serializing dataclasses using `flytekit` version >= v1.14.0, and you want to produce Protobuf `struct 
+literal` instead, you can set environment variable `FLYTE_USE_OLD_DC_FORMAT` to `true`.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:24]: no need to insert a new line, I think it breaks the code format

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 12]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
...
+
 :::{important}
-If you're using Flytekit version below v1.11.1, you will need to add `from dataclasses_json import dataclass_json` to your imports and decorate your dataclass with `@dataclass_json`.
+If you're using Flytekit version < v1.11.1, you will need to add `from dataclasses_json import dataclass_json` to your imports and decorate your dataclass with `@dataclass_json`.
+:::
+
+:::{important}
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:30]:
```suggestion
```

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 13]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
...
+Flytekit version < v1.14.0 will produce protobuf struct literal for dataclasses.
+
+Flytekit version >= v1.14.0 will produce msgpack bytes literal for dataclasses.
+
+If you're using Flytekit version >= v1.14.0 and you want to produce protobuf struct literal for dataclasses, you can
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:35]:
```suggestion
```

File: docs/user_guide/data_types_and_io/dataclass.md [hunk 14]
```diff
@@ -11,8 +11,31 @@ When you've multiple values that you want to send across Flyte entities, you can
...
+If you're using Flytekit version < v1.11.1, you will need to add `from dataclasses_json import dataclass_json` to your imports and decorate your dataclass with `@dataclass_json`.
+:::
+
+:::{important}
+Flytekit version < v1.14.0 will produce protobuf struct literal for dataclasses.
+
+Flytekit version >= v1.14.0 will produce msgpack bytes literal for dataclasses.
+
+If you're using Flytekit version >= v1.14.0 and you want to produce protobuf struct literal for dataclasses, you can 
+set environment variable  `FLYTE_USE_OLD_DC_FORMAT` to `true`.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/dataclass.md:36]:
```suggestion
```
This was already mentioned above

Also in the readthedocs build, you can see there are two `important` blocks nested

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
+(pydantic_basemodel)=
+
+# Pydantic BaseModel
+
+```{eval-rst}
+.. tags:: Basic
+```
+
+`flytekit` version >=1.14 supports natively the `JSON` format that Pydantic `BaseModel` produces,  enhancing the 
+interoperability of Pydantic BaseModels with the Flyte type system.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:10]: no need to insert new line

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+:::{important}
+Pydantic BaseModel V2 only works when you are using flytekit version >= v1.14.0.
+:::
+
+With the 1.14 release, `flytekit` adopted `MessagePack` as the serialization format for Pydantic `BaseModel`, 
+overcoming a major limitation of serialization into a JSON string within a Protobuf `struct` datatype like the previous versions do:
+
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:18]:
```suggestion
```

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+to store `int` types, Protobuf's `struct` converts them to `float`, forcing users to write boilerplate code to work around this issue.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:19]: I think new line after a colon (unless a list) looks strange

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+By default, `flytekit >= 1.14` will produce `msgpack` bytes literals when serializing dataclasses, preserving the types defined in your `BaseModel` class.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:21]:
```suggestion
```

File: docs/user_guide/data_types_and_io/pydantic_basemodel.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+:::{important}
+If you're serializing dataclasses using `flytekit` version >= v1.14.0 and you want to produce Protobuf `struct literal` instead, you can set environment variable `FLYTE_USE_OLD_DC_FORMAT` to `true`.
```
Comment (davidmirror-ops) [inline docs/user_guide/data_types_and_io/pydantic_basemodel.md:24]:
```suggestion
By default, `flytekit >= 1.14` will produce `msgpack` bytes literals when serializing, preserving the types defined in your `BaseModel` class.
If you're serializing `BaseModel` using `flytekit` version >= v1.14.0 and you want to produce Protobuf `struct literal` instead, you can set environment variable `FLYTE_USE_OLD_DC_FORMAT` to `true`.
```

Consolidating on the `important` block

**Comment by davidmirror-ops** [comment]
@Future-Outlier thank you for your patience. A few more suggestions, plus please make sure there are no nested `important` blocks.

File: docs/user_guide/data_types_and_io/dataclass.md (hunk 8 above)
Comment (Future-Outlier) [reply docs/user_guide/data_types_and_io/dataclass.md:17]: I think insert a new line is more readable

File: docs/user_guide/data_types_and_io/dataclass.md (hunk 7 above)
Comment (Future-Outlier) [reply docs/user_guide/data_types_and_io/dataclass.md:16]: I think new line is more readable
image

**Review by davidmirror-ops — APPROVED** [review]
Thank you

**Pull request merged**